pkg/apis/eventing/v1alpha1/channel_types.go

```diff
@@ -0,0 +1,172 @@
+var _ apis.Validatable = (*Channel)(nil)
+var _ apis.Defaultable = (*Channel)(nil)
+var _ apis.Immutable = (*Channel)(nil)
+var _ runtime.Object = (*Channel)(nil)
+var _ webhook.GenericCRD = (*Channel)(nil)
+
+// ChannelSpec specifies the Provisioner backing a channel and the configuration
+// arguments for a Channel.
+type ChannelSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Provisioner defines the name of the Provisioner backing this channel.
+	// TODO: +optional If missing, a default Provisioner may be selected for the Channel.
+	Provisioner *ProvisionerReference `json:"provisioner,omitempty"`
+
+	// Arguments defines the arguments to pass to the Provisioner which provisions
+	// this Channel.
+	// +optional
+	Arguments *runtime.RawExtension `json:"arguments,omitempty"`
+
+	// Subscribers is a list of the Subscribers to this channel. This is filled in
+	// by the Subscriptions controller. Users should not mutate this field.
+	Subscribers []ChannelSubscriberSpec `json:"subscribers,omitempty"`
+}
+
+// ChannelSubscriberSpec defines a single subscriber to a Channel. At least one
+// of Call or Result must be present.
+type ChannelSubscriberSpec struct {
+	// Call is an optional reference to a function for processing events.
+	// Events from the From channel will be delivered here and replies
+	// are optionally handled by Result.
+	// +optional
+	Call *Callable `json:"call,omitempty"`
+
+	// Result optionally specifies how to handle events received from the Call
+	// target.
+	// +optional
+	Result *ResultStrategy `json:"result,omitempty"`
+}
+
+// ChannelStatus represents the current state of a Channel.
+type ChannelStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Channel.
+	// It corresponds to the Channel's generation, which is updated on mutation by
+	// the API Server.
+	//TODO The above comment is only true once
+	// https://github.com/kubernetes/kubernetes/issues/58778 is fixed.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// DomainInternal holds the top-level domain that will distribute traffic
+	// over the provided targets from inside the cluster. It generally has the
+	// form {channel}.{namespace}.svc.cluster.local
+	//TODO move this to a struct that can be embedded similar to ObjectMeta and
+	// TypeMeta.
+	// +optional
+	DomainInternal string `json:"domainInternal,omitempty"`
+
+	// Represents the latest available observations of a channel's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []ChannelCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```

Comment:
you will want to update to the style of https://github.com/knative/eventing/pull/434

Reply:
Let's get #434 in first, reviewing now :)

Reply:
Updated to use duck Conditions.